### PR TITLE
extract capabilities from policies

### PR DESCRIPTION
**Pull request by mmandlis**

### CODE REVIEW

File: java/arcs/core/policy/Policy.kt [hunk 1]
```diff
@@ -12,6 +12,9 @@
 package arcs.core.policy
 
 import arcs.core.data.Annotation
+import arcs.core.data.Capabilities
+import arcs.core.data.Capability
+import java.util.concurrent.TimeUnit
```
Comment (csilvestrini) [inline java/arcs/core/policy/Policy.kt:17]: I don't think you can use java.util in core code, it's not multiplatform..

File: javatests/arcs/core/policy/PolicyTest.kt [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        assertThat(
+            capabilities[0].containsAll(
+                Capabilities(listOf(Persistence.ON_DISK, Encryption(true), Ttl.Minutes(120)))
+            )
+        )
+        assertThat(
+            capabilities[0].containsAll(
+                Capabilities(listOf(Persistence.IN_MEMORY, Ttl.Minutes(120)))
+            )
+        )
```
Comment (csilvestrini) [inline javatests/arcs/core/policy/PolicyTest.kt:43]: This seems wrong. Should the second one be `capabilities[1]`?

Also, you need to follow up `assertThat(...)` with something (`.containsAll`? `.isTrue`?)

Also, I think this whole thing should actually be:

```
assertThat(capabilities).containsExactly(
    Capabilities(listOf(...)),
    Capabilities(listOf(...))
).inOrder()
```

File: java/arcs/core/policy/Policy.kt (hunk 1 above)
Comment (mmandlis) [reply java/arcs/core/policy/Policy.kt]: thanks, removed

File: javatests/arcs/core/policy/PolicyTest.kt (hunk 2 above)
Comment (mmandlis) [reply javatests/arcs/core/policy/PolicyTest.kt]: oh my, this uncovered a lot of problems, including missing `isEquivalent` method in KT and broken `isEquivalent` method in TS. fixed, added tests. 
Thanks for the catch!

**Comment by mmandlis** [comment]
Cameron, thanks for the review.
That changes ended up being potentially larger than expected, so you might want to take another look.

PS: I'm ooo starting for the rest of the weeks, so if all checks pass sooner than you reply, i might merge, but will address comments, once i'm back!

File: java/arcs/core/data/Capabilities.kt [hunk 3]
```diff
@@ -65,6 +61,14 @@ class Capabilities(capabilities: List<Capability> = emptyList()) {
         return other.ranges.all { otherRange -> contains(otherRange) }
     }
 
+    fun isEquivalent(other: Capabilities): Boolean {
+        return ranges.size == other.ranges.size && other.ranges.all { hasEquivalent(it) }
+    }
+
+    fun hasEquivalent(capability: Capability): Boolean {
+        return ranges.find { it.isCompatible(capability) && it.isEquivalent(capability) } != null
```
Comment (csilvestrini) [inline java/arcs/core/data/Capabilities.kt:69]: Possibly you want `ranges.any { ... }` instead?

File: java/arcs/core/data/Capability.kt [hunk 4]
```diff
@@ -39,6 +39,17 @@ sealed class Capability(val tag: String) {
         }
     }
 
+    fun getRealTag(): String {
```
Comment (csilvestrini) [inline java/arcs/core/data/Capability.kt:42]: KDoc for this? (What's a "real tag"?)

File: javatests/arcs/core/data/CapabilitiesTest.kt [hunk 5]
```diff
@@ -24,9 +26,9 @@ class CapabilitiesTest {
     fun capabilities_empty() {
         assertThat(Capabilities().isEmpty).isTrue()
         assertThat(Capabilities.fromAnnotations(emptyList<Annotation>()).isEmpty).isTrue()
-        assertThat(Capabilities(Capability.Persistence.ON_DISK).isEmpty)
+        assertThat(Capabilities(Persistence.ON_DISK).isEmpty)
             .isFalse()
-        assertThat(Capabilities(listOf(Capability.Persistence.ON_DISK)).isEmpty)
+        assertThat(Capabilities(listOf(Persistence.ON_DISK)).isEmpty)
             .isFalse()
```
Comment (csilvestrini) [inline javatests/arcs/core/data/CapabilitiesTest.kt:32]: It looks like these fit on one line now

File: javatests/arcs/core/data/CapabilitiesTest.kt [hunk 6]
```diff
@@ -152,4 +154,68 @@ class CapabilitiesTest {
             )
         ).isFalse()
     }
+    @Test
+    fun capabilities_isEquivalent() {
+        val capabilities = Capabilities(
+            listOf(
+                Capability.Range(Ttl.Days(10), Ttl.Days(2))
+            )
```
Comment (csilvestrini) [inline javatests/arcs/core/data/CapabilitiesTest.kt:162]: One line?

File: javatests/arcs/core/policy/PolicyTest.kt (hunk 2 above)
Comment (csilvestrini) [reply javatests/arcs/core/policy/PolicyTest.kt]: Hooray for code reviews!

File: java/arcs/core/data/Capabilities.kt (hunk 3 above)
Comment (mmandlis) [reply java/arcs/core/data/Capabilities.kt]: right, thanks! done!

File: javatests/arcs/core/policy/PolicyTest.kt (hunk 2 above)
Comment (mmandlis) [reply javatests/arcs/core/policy/PolicyTest.kt]: oh yeah!! 👍 💯

File: java/arcs/core/data/Capability.kt (hunk 4 above)
Comment (mmandlis) [reply java/arcs/core/data/Capability.kt:46]: done (i'm not a huge fan of this name, but couldn't think of anything better).

File: javatests/arcs/core/data/CapabilitiesTest.kt (hunk 5 above)
Comment (mmandlis) [reply javatests/arcs/core/data/CapabilitiesTest.kt]: done

File: javatests/arcs/core/data/CapabilitiesTest.kt (hunk 6 above)
Comment (mmandlis) [reply javatests/arcs/core/data/CapabilitiesTest.kt]: done

**Pull request merged**